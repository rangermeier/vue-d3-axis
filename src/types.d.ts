type Scaler<T> = (x: T) => number

interface BandedScale<T> {
  (x: T): number
  domain: () => T[]
  range: () => number[]
  copy: () => BandedScale<T>
  bandwidth: () => number
  round: () => boolean
}


interface TickScale<T> {
  (x: T): number
  domain: () => T[]
  range: () => number[]
  ticks: (count: number) => T[]
  tickFormat: (count: number, fmt?: string | null | undefined) => (val: T) => string
  copy: () => TickScale<T>
  round: () => boolean
}


type Orients = 'TOP' | 'RIGHT' | 'BOTTOM' | 'LEFT'

interface AxisStyle {
  orient: Orients
  tickSizeInner: number
  tickSizeOuter: number
  tickPadding: number
  strokeWidth: number
  strokeColor: string
  tickFont: string
  tickFontSize: number
}


interface PropsForAxis<T> {
  range: number[]
  values: T[]
  position: Scaler<T>
  format: (d: T) => string
  styles?: Partial<AxisStyle>
}

interface LineProps {
  x1: number
  y1: number
  x2: number
  y2: number
  stroke: string
}

interface TextProps {
  fill: string
  dy: string
  x: number
  y: number
}
