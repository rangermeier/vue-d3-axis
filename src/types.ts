export type Scaler<T> = (x: T) => number

export interface BandedScale<T> {
  (x: T): number
  domain: () => T[]
  range: () => number[]
  copy: () => BandedScale<T>
  bandwidth: () => number
  round: () => boolean
}


export interface TickScale<T> {
  (x: T): number
  domain: () => T[]
  range: () => number[]
  ticks: (count: number) => T[]
  tickFormat: (count: number, fmt?: string | null | undefined) => (val: T) => string
  copy: () => TickScale<T>
  round: () => boolean
}


export type Orients = 'TOP' | 'RIGHT' | 'BOTTOM' | 'LEFT'

export interface AxisStyle {
  orient: Orients
  tickSizeInner: number
  tickSizeOuter: number
  tickPadding: number
  strokeWidth: number
  strokeColor: string
  tickFont: string
  tickFontSize: number
}


export interface PropsForAxis<T> {
  range: number[]
  values: T[]
  position: Scaler<T>
  format: (d: T) => string
}
