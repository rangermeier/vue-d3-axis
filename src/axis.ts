import { h } from 'vue'

function translateX<T>(scale0: Scaler<T>, scale1: Scaler<T>, d: T): string {
  const x = scale0(d)
  return `translate(${isFinite(x) ? x : scale1(d)},0)`
}

function translateY<T>(scale0: Scaler<T>, scale1: Scaler<T>, d: T): string {
  const y = scale0(d)
  return `translate(0,${isFinite(y) ? y : scale1(d)})`
}

export const TOP = 'TOP'
export const RIGHT = 'RIGHT'
export const BOTTOM = 'BOTTOM'
export const LEFT = 'LEFT'

const defaultAxisStyle: AxisStyle = {
  orient: BOTTOM,
  tickSizeInner: 6,
  tickSizeOuter: 6,
  tickPadding: 3,
  strokeWidth: 1,
  strokeColor: 'currentColor',
  tickFont: 'sans-serif',
  tickFontSize: 10
}

export const Axis = <T>(props: PropsForAxis<T>) => {
  const axisStyle = { ...defaultAxisStyle, ...props.styles }
  const {
    orient,
    tickSizeInner,
    tickPadding,
    tickSizeOuter,
    strokeWidth,
    strokeColor,
    tickFont,
    tickFontSize
  } = axisStyle

  const transform = orient === TOP || orient === BOTTOM
    ? translateX
    : translateY
  const tickTransformer = (d: T): string => transform(props.position, props.position, d)

  const k = orient === TOP || orient === LEFT ? -1 : 1
  const isRight = orient === RIGHT
  const isLeft = orient === LEFT
  const isTop = orient === TOP
  const isBottom = orient === BOTTOM
  const isHorizontal = isRight || isLeft
  const x = isHorizontal ? 'x' : 'y'
  const y = isHorizontal ? 'y' : 'x'

  const halfWidth = strokeWidth / 2
  const range0 = props.range[0] + halfWidth
  const range1 = props.range[props.range.length - 1] + halfWidth

  const spacing = Math.max(tickSizeInner, 0) + tickPadding

  return h(
    'g',
    {
      fill: 'none',
      fontSize: tickFontSize,
      fontFamily: tickFont,
      textAnchor: isRight ? 'start' : isLeft ? 'end' : 'middle',
      strokeWidth: strokeWidth
    },
    [
      h('path', {
        stroke: strokeColor,
        d: isHorizontal
          ? `M${k * tickSizeOuter},${range0}H${halfWidth}V${range1}H${k * tickSizeOuter}`
          : `M${range0},${k * tickSizeOuter}V${halfWidth}H${range1}V${k * tickSizeOuter}`
      }),
      props.values.map((v, idx) => {
        const lineProps: Partial<LineProps> = { stroke: strokeColor }
        lineProps[`${x}2`] = k * tickSizeInner
        lineProps[`${y}1`] = halfWidth
        lineProps[`${y}2`] = halfWidth

        const textProps: Partial<TextProps> = {
          fill: strokeColor,
          dy: isTop ? '0em' : isBottom ? '0.71em' : '0.32em'
        }
        textProps[`${x}`] = k * spacing
        textProps[`${y}`] = halfWidth
        return h('g', {
          key: `tick-${idx}`,
          opacity: 1,
          transform: tickTransformer(v)
        }, [
          h('line', lineProps),
          h('text', textProps, props.format(v)),
        ])
      })
    ]
  )
}
