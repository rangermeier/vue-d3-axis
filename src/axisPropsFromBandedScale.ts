import { BandedScale, PropsForAxis } from './types'

export function axisPropsFromBandedScale<T>(
  scale: BandedScale<T>,
): PropsForAxis<T> {
  const range = scale.range()
  const values = scale.domain()
  const format = (d: T): string => String(d)

  let offset = scale.bandwidth() / 2
  const scaleCopy = scale.copy()
  if (scaleCopy.round()) offset = Math.round(offset)
  const position = (d: T): number => scaleCopy(d) + offset

  return { range, values, format, position }
}
