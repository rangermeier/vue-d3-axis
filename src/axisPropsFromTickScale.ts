export function axisPropsFromTickScale<T>(
  scale: TickScale<T>, tickCount: number
): PropsForAxis<T> {
  const range = scale.range()
  const values = scale.ticks(tickCount)
  const format = scale.tickFormat(tickCount)

  const position = scale.copy()

  return { range, values, format, position }
}
