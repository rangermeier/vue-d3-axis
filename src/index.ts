export { axisPropsFromTickScale } from './axisPropsFromTickScale'
export { axisPropsFromBandedScale } from './axisPropsFromBandedScale'
export { Axis, TOP, RIGHT, BOTTOM, LEFT } from './axis'
