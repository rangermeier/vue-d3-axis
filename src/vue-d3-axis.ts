export { default as axisPropsFromTickScale } from './axisPropsFromTickScale'
export { default as axisPropsFromBandedScale } from './axisPropsFromBandedScale'
export { default as Axis, TOP, RIGHT, BOTTOM, LEFT } from './Axis'
export { TickScale, BandedScale, Scaler, Orients, AxisStyle } from './types'
