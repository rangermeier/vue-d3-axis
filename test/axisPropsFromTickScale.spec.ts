import { expect, test } from 'vitest'
import {scaleLinear} from 'd3-scale'
import { axisPropsFromTickScale } from '../src/axisPropsFromTickScale'

test('Creates axis props', () => {
  const scale = scaleLinear().domain([0, 100]).range([0, 500])
  const props = axisPropsFromTickScale(scale, 10)
  expect(props).toMatchSnapshot()

  expect(props.format(50)).toEqual('50')
  expect(props.position(50)).toEqual(250)
})
