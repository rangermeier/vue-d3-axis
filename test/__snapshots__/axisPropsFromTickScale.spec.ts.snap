// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Creates axis props 1`] = `
{
  "format": [Function],
  "position": [Function],
  "range": [
    0,
    500,
  ],
  "values": [
    0,
    10,
    20,
    30,
    40,
    50,
    60,
    70,
    80,
    90,
    100,
  ],
}
`;
